### PR TITLE
feat(ui): allow adding storage to existing datastores

### PR DESCRIPTION
**Pull request by Caleb-Ellis**
## Done

- Added "Add to existing datastore" to VMFS6 bulk actions

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the storage tab of a machine with at least 2 disks, and is in Ready or Allocated state. (If you set the MAAS to karura you can find one at http://0.0.0.0:8400/MAAS/r/machine/8qcbef/storage).
- Change the storage layout to VMFS6, which will automatically partition and use one of the disks
- Partition the other disk into 3 unformatted partitions
- Select one of the partitions and create a datastore from it
- Select both of the remaining partitions and click "Add to existing datastore"
- Check that the total size is correct
- Check that you can choose from the 2 newly created datastores, and the mount point updates correctly
- Check that you can successfully add storage to a datastore

## Fixes

Fixes canonical-web-and-design/MAAS-squad#2172

### CODE REVIEW

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx [hunk 1]
```diff
@@ -38,21 +38,12 @@ type Props = {
...
-      disk.partitions.forEach((partition) => {
-        if (isDatastore(partition.filesystem)) {
-          count += 1;
-        }
-      });
-    }
-    return count;
-  }, 0);
+  const datastoresCount = disks.reduce<number>(
+    (count, disk) => (isDatastore(disk.filesystem) ? count + 1 : count),
```
Comment (Caleb-Ellis) [inline ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx:44]: Driveby fix. We actually only want to count the disks, not the partitions. Also the api starts at 1, so I've updated it to start at 1 here too

File: ui/src/app/store/machine/utils/storage.ts [hunk 2]
```diff
@@ -317,7 +347,7 @@ export const isCacheSet = (disk: Disk | null): boolean =>
  * @param fs - the filesystem to check.
  * @returns whether the filesystem is a VMFS6 datastore
  */
-export const isDatastore = (fs: Filesystem | null): fs is Filesystem =>
+export const isDatastore = (fs: Filesystem | null): boolean =>
```
Comment (Caleb-Ellis) [inline ui/src/app/store/machine/utils/storage.ts:350]: Driveby fix. `fs is Filesystem` is incorrect - just because something isn't a datastore doesn't mean it's also not a filesystem.

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx:44]: Nice. Might be good to leave a comment here about starting at 1 like you did in the test.

**Review by huwshimi — APPROVED** [review]

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateDatastore/CreateDatastore.tsx:44]: Ah, thought I did already. Adding now
